include/boost/gil/extension/dynamic_image/apply_operation.hpp

```diff
@@ -10,47 +10,47 @@
 
 #include <boost/gil/detail/mp11.hpp>
 
-#include <boost/variant/apply_visitor.hpp>
+#include <boost/variant2/variant.hpp>
 
 namespace boost { namespace gil {
 
 /// \ingroup Variant
 /// \brief Invokes a generic mutable operation (represented as a unary function object) on a variant
-template <typename Types, typename UnaryOp>
+template <typename ...Types, typename UnaryOp>
 BOOST_FORCEINLINE
-auto apply_operation(variant<Types>& arg, UnaryOp op)
+auto apply_operation(variant2::variant<Types...>& arg, UnaryOp op)
 #if defined(BOOST_NO_CXX14_DECLTYPE_AUTO) || defined(BOOST_NO_CXX11_DECLTYPE_N3276)
     -> typename UnaryOp::result_type
```

Comment:
Can we do `decltype(arg(op))` here? Ditto for other overloads.

Reply:
@simmplecoder Only as `#else` case, I think. However, I'd suggest to do it in separate PR.

Reply:
@mloskot , `decltype` was added in C++11 [cppreference](https://en.cppreference.com/w/cpp/language/decltype). I can test that out, but only on weekends probably.

Reply:
@simmplecoder  Yes, but there is `BOOST_NO_CXX11_DECLTYPE` vs `BOOST_NO_CXX11_DECLTYPE_N3276` which I'm not sure about

Reply:
My humble opinion is that `apply_operation` should be marked `[[deprecated]]` and provided only for backward compatibility. Just use `visit` in modern code (no limit on arity, etc...)

Reply:
I agree. There is lots of legacy constructs that can be deprecated and avoided now. 
It will take another refactoring overhaul, or two. I just didn't have time to move to that.

Reply:
I agree as well. I believe those apply functions also don't catch easy errors, like when one forgets `gil::view` or something like that, and due to upcast it is unclear why the functor cannot be called.

Reply:
@mloskot , I believe boost variant2 itself might depend on `BOOST_NO_CXX11_DECLTYPE_N3276`. I will try to find out if it is really the case.

Reply:
I made a TODO from the whole idea https://github.com/boostorg/gil/issues/482